inmaputil/download.go

```diff
@@ -0,0 +1,122 @@
+// If so, it downloads it and checks if the file needs to be decompressed.
+// It either returns the path of the Shp file or decompresses it and returns the path 
+// to the decompressed Shp file.
+// The second argument allows to specify a channel where logs will be written
+func download(path string, c chan string) string {
+	// Check if local file exists. If it does, return the give path.
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
+		return path
+	} 
+
+	// If the input doesn't start with these prefixes, then return the input.
+	if !strings.HasPrefix(path, "http://") && !strings.HasPrefix(path, "https://") {
+		return path
+	}
+
+	// Preparing a temporary directory for the downloads.
+	tempDir, err := ioutil.TempDir("","inmap");
+	if err != nil {
+		panic(fmt.Errorf("inmaputil: failed creating temporary download directory: %v", err))
+		return path
+	}
+	
+	// Start the downloading client.
+	client := grab.NewClient()
+	req, err := grab.NewRequest(tempDir, path)
+	if err != nil {
+		panic(fmt.Errorf("Failed creating downloading Request."))
+	}
+
+	c <- fmt.Sprintf("Downloading ... %v\n" , req.URL())
+	resp := client.Do(req)
+
+	// Start a ticker to print out progress.
+	t := time.NewTicker(1000 * time.Millisecond)
+	defer t.Stop()
+
+Loop:
+	for {
+		select {
+			case <-t.C:
+				c <- fmt.Sprintf("   transferred %v / %v bytes", resp.BytesComplete(), resp.Size)
+			case <- resp.Done:
+				break Loop
+		}
+	}
+
+	if err := resp.Err(); err != nil {
+		c <- fmt.Sprintf("Download failed: %v", err)
+		return path
+	}
+
+	c <- fmt.Sprintf("Download saved to ./%v ", resp.Filename)
+	c <- ("Decompressing...")
+
+	// Check if the downloaded file is compressed.
+	// If so, decompress it according to its extension.
+	switch filepath.Ext(resp.Filename) {
+	case ".zip":
+		err := archiver.Zip.Open(resp.Filename, resp.Filename + "_extracted")
+		if err != nil {
+			return resp.Filename
+		}
+	case ".gz":
+		err := archiver.TarGz.Open(resp.Filename, resp.Filename + "_extracted")
+		if err != nil {
+			return resp.Filename
+		}
+	case ".tar":
+		err := archiver.Tar.Open(resp.Filename, resp.Filename + "_extracted")
+		if err != nil {
+			return resp.Filename
+		}
+	default:
+		// If not compressed, return the path to the downloaded file
+		return resp.Filename
+	}
+
+
+	// choose the Shp file in the downloaded path.
+	// If Shp file is not found, try going in the first directory
+
+	foundShp := "";
+
+	err = filepath.Walk(resp.Filename + "_extracted/", func(path string, info os.FileInfo,
+		err error) error {
+			if err != nil {
+				panic(fmt.Errorf("There was a problem reading extracted file. "))
+			}
+			if !info.IsDir() && strings.HasSuffix(path, ".shp") {
+				// Found a Shp file, returns it.
+				foundShp = path;
+			}
+			return nil
+	})
+
+	// We've walked the directory but found no Shp file
+	// So we return the downloaded path.
+	if err != nil {
+		return resp.Filename
+	} else {
+		return foundShp;
+	}
+
+	close(c)
```

Comment:
Probably delete this line; the caller may want to reuse the channel somewhere else.